### PR TITLE
crimson/os: generalize AlienStore beyond BlueStore.

### PR DESCRIPTION
**Pull request by rzarzynski**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/os/futurized_store.cc [hunk 1]
```diff
@@ -14,9 +14,13 @@ FuturizedStore::create(const std::string& type,
   if (type == "memstore") {
     return std::make_unique<crimson::os::CyanStore>(data);
   } else if (type == "bluestore") {
-    return std::make_unique<crimson::os::AlienStore>(data, values, alien);
+    return std::make_unique<crimson::os::AlienStore>(
+      type, data, values, alien);
   } else if (type == "seastore") {
     return crimson::os::seastore::make_seastore(data, values);
+  } else if (type.rfind("alienstore+", 0) == 0) {
```
Comment (rzarzynski) [inline src/crimson/os/futurized_store.cc:21]: `std::string::starts_with` is available since C+20, sorry.

File: src/os/ObjectStore.h [hunk 2]
```diff
@@ -75,15 +75,21 @@ class ObjectStore {
    *
    * @param type type of store. This is a std::string from the configuration file.
    * @param data path (or other descriptor) for data
-   * @param journal path (or other descriptor) for journal (optional)
+   * @param journal path (or ,other descriptor) for journal (optional)
```
Comment (tchaikov) [inline src/os/ObjectStore.h:78]: might want to drop this change?

File: src/common/options/global.yaml.in [hunk 3]
```diff
@@ -3622,6 +3622,9 @@ options:
   - memstore
   - kstore
   - seastore
+  - alienstore+bluestore
+  - alienstore+memstore
+  - alienstore+random
```
Comment (tchaikov) [inline src/common/options/global.yaml.in:3627]: what's the use case of "random"?

File: src/crimson/os/futurized_store.cc (hunk 1 above)
Comment (tchaikov) [reply src/crimson/os/futurized_store.cc:21]: why do we need to add new store types? i think the plain "bluestore", "memstore" would suffice. instead, how about just adding "cyanstore" to allowed values of "osd_objectstore"?

File: src/os/ObjectStore.h (hunk 2 above)
Comment (rzarzynski) [reply src/os/ObjectStore.h:78]: Oops, sure!

File: src/common/options/global.yaml.in (hunk 3 above)
Comment (rzarzynski) [reply src/common/options/global.yaml.in:3627]: The sole I can imagine is a teuthology testing.

File: src/crimson/os/futurized_store.cc (hunk 1 above)
Comment (rzarzynski) [reply src/crimson/os/futurized_store.cc:21]: I think the problem is in legacy. We already started expecting `CyanStore` when `memstore` is passed (`--memstore` of `vstart.sh`).

Comment (tchaikov) [reply src/crimson/os/futurized_store.cc:21]: i'd prefer fixing this at a higher level -- just add a `--cyanstore` option. and update `doc/dev/crimson/crimson.rst` accordingly. i can do it.

File: src/common/options/global.yaml.in (hunk 3 above)
Comment (tchaikov) [reply src/common/options/global.yaml.in:3627]: i'd use a `$` facet to do this.

File: src/crimson/os/futurized_store.cc (hunk 1 above)
Comment (rzarzynski) [reply src/crimson/os/futurized_store.cc]: Fixed and repushed!

**Review by tchaikov — APPROVED** [review]

**Review by liu-chunmei — APPROVED** [review]

**Review by markhpc — APPROVED** [review]
@rzarzynski This is great! :+1:

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-07-22_12:07:24-rados-wip-kefu-testing-2021-07-22-1810-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-23_01:40:03-rados-wip-kefu-testing-2021-07-21-2006-distro-basic-smithi/

File: src/crimson/os/futurized_store.cc (hunk 1 above)
Comment (tchaikov) [reply src/crimson/os/futurized_store.cc]: > i'd prefer fixing this at a higher level -- just add a `--cyanstore` option. and update `doc/dev/crimson/crimson.rst` accordingly. i can do it.

#42476

**Pull request merged**